### DISCUSSION TITLE
multiple middleware

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#63449] [Ideas]
multiple middleware with app directory like page.tsx we have middleware.tsx for every route

**Comment by icyJoseph** [comment]
This was the initial design for middleware, all the way back in Pages Router, version 12, but it was reworked to only root middleware, you can read more about it here:

- https://github.com/vercel/next.js/pull/36772

**Reply by Falci** [reply]
Do you know why this decision?

[Docs](https://nextjs.org/docs/app/building-your-application/routing/middleware#convention) says:
> Note: While only one middleware.ts file is supported per project, you can still organize your middleware logic modularly. Break out middleware functionalities into separate .ts or .js files and import them into your main middleware.ts file. This allows for cleaner management of route-specific middleware, aggregated in the middleware.ts for centralized control. By enforcing a single middleware file, it simplifies configuration, prevents potential conflicts, and optimizes performance by avoiding multiple middleware layers.

"Break out middleware functionalities into separate .ts or .js files" is ok, but "import them into your main middleware.ts file" should be next.js' job, no?

**Reply by MrOxMasTer** [reply]
The problem is that middleware is activated for each route. Of course, you can make a modular approach, or rather, embed middleware into each other.But then, instead of the simple syntax of the middleware file and a separate array with designated paths (which in most cases is simply not useful), we get porridge from middleware and from endless checking conditions at the beginning of the path and it does not look as beautiful as in a separate folder (app routing) or group. And when you have one middleware embedded in another, it's also not so clear at what stage which one will work or broke.

**Comment by MrOxMasTer** [comment]
> This was the initial design for middleware, all the way back in Pages Router, version 12, but it was reworked to only root middleware, you can read more about it here:
> 
> * [Deprecate nested Middleware in favor of root middleware #36772](https://github.com/vercel/next.js/pull/36772)

That's a terrible idea. Not only is one middleware simply not enough, but it is also not convenient.

For example, to set cookies I need to use either middleware for a couple of pages, which is one for the whole project, because I can't set cookies in server components, or use a client component, which makes no sense at all to make it a client component, except in this situation.
